lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,177 @@
+        end
+
+        exit_msg = "PowerShell process did not terminate in reasonable time"
+        Timeout.timeout(3) do
+          Puppet.debug "Awaiting PowerShell process termination..."
+          @exit_status = @ps_process.value
+        end
+
+        exit_msg = "PowerShell process exited: #{@exit_status}" if @exit_status
+        Puppet.debug(exit_msg)
+        @ps_process.terminate if @ps_process.alive?
+      end
+
+      private
+
+      def self.is_readable?(stream, timeout = 0.5)
+        read_ready = IO.select([stream], [], [], timeout)
+        read_ready && stream == read_ready[0][0]
+      end
+
+      def self.create_event(name, manual_reset = false, initial_state = false)
+        if Puppet::Util::Platform.windows?
+          handle = FFI::Pointer::NULL_HANDLE
+
+          FFI::Pointer.from_string_to_wide_string(name) do |name_ptr|
+            handle = CreateEventW(FFI::Pointer::NULL,
+              manual_reset ? 1 : FFI::WIN32_FALSE,
+              initial_state ? 1 : FFI::WIN32_FALSE,
+              name_ptr)
+
+            if handle == FFI::Pointer::NULL_HANDLE
+              msg = "Failed to create new event #{name}"
+              raise Puppet::Util::Windows::Error.new(msg)
+            end
+          end
+        end
+
+        handle
+      end
+
+      WAIT_ABANDONED = 0x00000080
+      WAIT_OBJECT_0 = 0x00000000
+      WAIT_TIMEOUT = 0x00000102
+      WAIT_FAILED = 0xFFFFFFFF
+
+      def self.wait_on(wait_object, timeout = 120 * 1000, wait_interval = 50)
+        waited = 0
+        while true
+          wait_result = Puppet::Util::Windows::Process::WaitForSingleObject(wait_object, wait_interval)
+          case wait_result
+          when WAIT_OBJECT_0
+            Puppet.debug "Wait object signaled"
+            return
+          when WAIT_TIMEOUT
+            waited += wait_interval
+            Puppet.debug "Waiting #{wait_interval} milliseconds (total #{waited})..."
+            if (waited > timeout)
+              msg = 'Catastropic failure: wait object timeout exceeded'
+              raise Puppet::Util::Windows::Error.new(msg)
+            end
+          # only applicable to mutexes - should never happen here
+          when WAIT_ABANDONED
+            msg = 'Catastropic failure: wait object in inconsistent state'
+            raise Puppet::Util::Windows::Error.new(msg)
+          when WAIT_FAILED
+            msg = 'Catastropic failure: waiting on object to be signaled'
+            raise Puppet::Util::Windows::Error.new(msg)
+          end
+        end
+      end
+
+      def write_stdin(input)
+        @stdin.puts(input)
+      rescue => e
+        msg = "Error writing STDIN / reading STDOUT: #{e}"
+        raise Puppet::Util::Windows::Error.new(msg)
+      end
+
+      def read_stdout
+        output = []
+        self.class.wait_on(@output_ready_event)
+
+        initially_readable = false
+        while self.class.is_readable?(@stdout, 0.1) do
```

Comment:
> line on 138 blocks until the @output_ready_event is signaled from the PowerShell side after it has executed and written all it's data

If powershell needs to write more than the named pipe buffer (I forget if it's 4 or 8K by default), then ruby and powershell will deadlock. That's because powershell will block writing to stdout waiting for ruby to drain the pipe, and ruby will wait for powershell to signal the ready event, which will never happen.